src/System.Private.CoreLib/shared/System/Collections/HashHelpers.cs

```diff
@@ -86,5 +89,33 @@ public static int ExpandPrime(int oldSize)
 
             return GetPrime(newSize);
         }
+
+        // Takes an index and distributes it in the range [0, range)
+        // e.g. 0 to range exclusive.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint FastRange(uint index, uint range)
+        {
+            // Sse42.IsSupported check is baked into the R2R image so while the value returned
+            // from this method will be different on both branches, it will remain stable during
+            // a process' lifetime even if it is recompiled by tiered compilation.
+            if (Sse42.IsSupported)
+            {
+                // Using fastrange from Daniel Lemire https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/
+                //
+                // While fastrange is "fair"; it expects an evenly distributed hashCode
+                // from [0,2^32) to produce a distibuted range. As many hashCodes in .NET
+                // are generated with the high entropy in the LSB we need to redistribute
+                // the hash; for this we use Crc32. These two together allow us to skip the
+                // more costly idiv instruction.
+
+                const uint CrcReflected = 0xEDB88320;
+                return (uint)((Sse42.Crc32(CrcReflected, index) * (ulong)range) >> 32);
```

Comment:
@mikedn @AndyAyersMS running the hashcode through Crc32 gets around the distribution issue with unknown hashcodes being high entropy in their LSB when using @lemire's fastrange. (`int` being an example here)

Though alas we can only do this when hardware accelerated as otherwise Crc32 is too slow; so fall back to modulo when its not supported in hardware.

Reply:
I can confirm from my experience that fastrange is very dangerous when used as a hash reducer. I tried it in the past and it resulted in massive collisions in common cases like sequential numbers, zip codes, etc. - i.e. stuff that has entropy in low bits. 

I did not consider hardware-accelerated mixer like Crc32 - cannot comment on that. It was not available at the time. It is an interesting idea.

My guess it may still regress cases like when dictionary is used as a sparse list and keys have clusters. Modulus, for good or bad, happens to preserve the locality.

Reply:
Need to reverse the bits to reverse the entropy; and then you get something closer to the behaviour of modulus with fastrange. Alas there still isn't a fast way to do that in hardware; and reversing the endianness still leaves the wrong order entropy per byte.

Re: locality https://github.com/dotnet/coreclr/pull/27149#discussion_r334342237 for Dictionary specifically it has a lesser impact; than for example Hashtable, as its only changing the locality of the smaller int[] array that contains the indexes and the larger Entry[] array would maintain the same very packed locality. But yes it would have an impact.

Reply:
One approach could be to use a bucket array that was of the correct type for the size (e.g. dictionary size 1-255 could be a `byte[]`, 256-65536 could be an `ushort[]`, etc)